src/firestore-adapter.ts

```diff
@@ -0,0 +1,858 @@
+import "firebase/compat/database";
+
+import firebase from "firebase/compat/app";
+
+import { CursorType, ICursor } from "./cursor";
+import {
+  DatabaseAdapterCallbackType,
+  DatabaseAdapterEvent as FirebaseAdapterEvent,
+  IDatabaseAdapter,
+  IDatabaseAdapterEvent,
+  SendCursorCallbackType,
+  SendOperationCallbackType,
+  UserIDType,
+} from "./database-adapter";
+import { EventEmitter, EventListenerType, IEventEmitter } from "./emitter";
+import {
+  ITextOperation,
+  TextOperation,
+  TextOperationType,
+} from "./text-operation";
+import * as Utils from "./utils";
+
+type FirebaseRefCallbackType = (
+  snapshot: firebase.database.DataSnapshot
+) => void;
+
+type FirebaseRefCallbackHookType = {
+  ref: firebase.database.Reference | firebase.database.Query;
+  eventType: firebase.database.EventType;
+  callback: FirebaseRefCallbackType;
+  context?: ThisType<FirestoreAdapter>;
+};
+
+type RevisionType = {
+  /** Author */
+  a: UserIDType;
+  /** Operation */
+  o: TextOperationType;
+};
+
+type RevisionHistoryType = {
+  [revisionId: string]: RevisionType;
+};
+
+export type FirebaseOperationDataType = RevisionType & {
+  /** Timestamp */
+  t: any;
+};
+
+export type FirebaseCursorDataType = {
+  /** Color of Cursor */
+  color: string;
+  /** Name of User */
+  name: string;
+  /** Position of Cursor/Selection */
+  cursor: CursorType;
+};
+
+// Based off ideas from http://www.zanopha.com/docs/elen.pdf
+const characters =
+  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+export interface IFirebaseAdapterEvent extends IDatabaseAdapterEvent {}
+
+/** Copy of the Operation and Revision ID just sent */
+interface ISentOperation {
+  /** Revision ID */
+  id: string;
+  /** Operation Sent to Server */
+  op: ITextOperation;
+}
+
+/** Parsed Operation data from JSON representation */
+interface IRevision {
+  /** Author */
+  author: UserIDType;
+  /** Operation */
+  operation: ITextOperation;
+}
+
+export class FirestoreAdapter implements IDatabaseAdapter {
+  protected _zombie: boolean;
+  protected _initialRevisions: boolean;
+  protected _ready: boolean;
+  protected _revision: number;
+  protected _sent: ISentOperation | null;
+  protected _checkpointRevision: number;
+  protected _userId: UserIDType | null;
+  protected _userColor: string | null;
+  protected _userName: string | null;
+  protected _userCursor: ICursor | null;
+  protected _pendingReceivedRevisions: RevisionHistoryType;
+  protected _emitter: IEventEmitter | null;
+  protected _document: ITextOperation | null;
+  protected _userRef: firebase.database.Reference | null;
+  protected _firestoreUserRef: firebase.firestore.DocumentReference | null;
+  protected _firestoreRef: firebase.firestore.DocumentReference | null;
```

Comment:
can we use `protected _firestoreUserRef: firebase.firestore.DocumentReference` and remove the null type as I can see that the `_firestoreUserRef` is always assigned in constructor and it will help us to avoid `this._firestoreUserRef!.somemethod`

Reply:
We are nulling it [here](https://github.com/interviewstreet/firepad-x/blob/b32673b54914b791e4e6b10f0870f27e6340f433/src/firestore-adapter.ts#L663)